sdks/python/apache_beam/internal/gcp/auth.py

```diff
@@ -57,77 +71,58 @@ def set_running_in_gce(worker_executing_project):
-    VM instance.
-
-    Args:
-      user_agent: string, The HTTP User-Agent to provide for this application.
-    """
-    super(_GCEMetadataCredentials, self).__init__(
-        None,  # access_token
-        None,  # client_id
-        None,  # client_secret
-        None,  # refresh_token
-        datetime.datetime(2010, 1, 1),  # token_expiry, set to time in past.
-        None,  # token_uri
-        user_agent)
-
-  @retry.with_exponential_backoff(
-      retry_filter=retry.retry_on_server_errors_and_timeout_filter)
-  def _refresh(self, http_request):
-    refresh_time = datetime.datetime.utcnow()
-    metadata_root = os.environ.get(
-        'GCE_METADATA_ROOT', 'metadata.google.internal')
-    token_url = ('http://{}/computeMetadata/v1/instance/service-accounts/'
-                 'default/token').format(metadata_root)
-    req = Request(token_url, headers={'Metadata-Flavor': 'Google'})
-    token_data = json.loads(urlopen(req, timeout=60).read().decode('utf-8'))
-    self.access_token = token_data['access_token']
-    self.token_expiry = (refresh_time +
-                         datetime.timedelta(seconds=token_data['expires_in']))
-
-
 def get_service_credentials():
   """For internal use only; no backwards-compatibility guarantees.
 
-  Get credentials to access Google services."""
-  user_agent = 'beam-python-sdk/1.0'
-  if is_running_in_gce:
-    # We are currently running as a GCE taskrunner worker.
-    #
-    # TODO(ccy): It's not entirely clear if these credentials are thread-safe.
-    # If so, we can cache these credentials to save the overhead of creating
-    # them again.
-    return _GCEMetadataCredentials(user_agent=user_agent)
-  else:
-    client_scopes = [
-        'https://www.googleapis.com/auth/bigquery',
-        'https://www.googleapis.com/auth/cloud-platform',
-        'https://www.googleapis.com/auth/devstorage.full_control',
-        'https://www.googleapis.com/auth/userinfo.email',
-        'https://www.googleapis.com/auth/datastore'
-    ]
-
-    try:
-      credentials = GoogleCredentials.get_application_default()
-      credentials = credentials.create_scoped(client_scopes)
-      logging.debug('Connecting using Google Application Default '
-                    'Credentials.')
-      return credentials
-    except Exception as e:
-      logging.warning(
-          'Unable to find default credentials to use: %s\n'
-          'Connecting anonymously.', e)
-      return None
+  Get credentials to access Google services.
+
+  Returns:
+    A ``oauth2client.client.OAuth2Credentials`` object or None if credentials
+    not found. Returned object is thread-safe.
+  """
+  return _Credentials.get_service_credentials()
+
+
+class _Credentials(object):
+  _credentials_lock = threading.Lock()
+  _credentials_init = False
+  _credentials = None
+
+  @classmethod
+  def get_service_credentials(cls):
+    if cls._credentials_init:
+      return cls._credentials
```

Comment:
Is it possible for credentials to expire?

Reply:
yes; they auto-refresh